cEP-0031.md

```diff
@@ -0,0 +1,254 @@
+    correct indentation in the program statements since that is crucial for
+    proper code style. Docstrings/comments, on the other hand, do not have any
+    such compulsion. So ideally docstrings should be ignored by the bear. coala
+    already uses documentation extraction algorithms in DocumentationStyleBear.
+    Using the existing algorithms and constructing ignore ranges out of the
+    extracted comments inside the IndentationBear, this issue can be solved.
+
+-   [coala/coala-bears#502][502]: MarkdownBear is not very helpful
+    On running MarkdownBear all we get is the result message 'The text does not
+    comply to the set style.'. This is not very helpful for most users. We will
+    be using a plugin to solve this issue.
+
+-   [coala/coala-bears#1897][1897]: Show PEP8 error description
+    On running PEP8Bear the default message is 'The code does not comply to
+    PEP8.' which is not very helpful for the user. This issue will be solved
+    by implementing upstream changes to autopep8.
+
+-   [hhatto/autopep8#227][227]: Create diff for a specific pep8 issue
+    User should have a choice of selecting the issue he wants to generate a fix
+    for. Although the issue is a part of the autopep8, we will be solving
+    this as a part of upstream changes to autopep8.
+
+Along with fixing these issues, new bears including `OutdatedDependencyBear`,
+`FileExistsBear`, `RegexLintBear`, `FileModeBear` and `RequirementsCheckBear`
+will be implemented.
+
+## Implementation of new bears
+
+### 1. OutdatedDependencyBear
+
+Issue: [coala/coala-bears#2445][2445]
+
+New functions to the existing package manager would be the best way of
+implementing this.
+
+- For pip requirements, we could use the PyPI JSON API which provides the
+    version info by specifying the `package_name`.
+
+```py
+class PipRequirement(PackageRequirement):
+    def get_version_info(self):
+        # This would get the package info in json format
+        url = 'https://pypi.python.org/pypi/%s/json' % (self.package,)
+
+        # Get the required data
+        data = json.load()
+        versions = data['releases'].keys().sort(key = StrictVersion)
+        return versions
+```
+
+- For npm dependencies, we can use the npm cli command `npm outdated` which
+    produces an output as below:
+
+```
+$ npm outdated
+Package      Current   Wanted   Latest  Location
+glob          5.0.15   5.0.15    6.0.1  test-outdated-output
+nothingness    0.0.3      git      git  test-outdated-output
+npm            3.5.1    3.5.2    3.5.1  test-outdated-output
+local-dev      0.0.3   linked   linked  test-outdated-output
+once           1.3.2    1.3.3    1.3.3  test-outdated-output
+```
+
+The output can be directly parsed and can be used to get the list of all
+outdated dependencies.
+
+### 2. FileExistBear
+
+Issue: [coala/coala-bears#527][527]
+
+The bear will be responsible to check if the files exist in the project
```

Comment:
Why did you decide to not implement it like I suggested at

https://github.com/coala/coala-bears/issues/527#issuecomment-284317117

Reply:
Let me know if I'm wrong:

You want to check for the files' existence through the `files` itself, and moreover, they should have explicit mentions (no globs) and produce an error message if the files are not present.

Also, why no globs?

Reply:
The 'no globs' bit was because it is easier, and achieves almost the same result.  Handling of globs would be better, but needs more thought put into the design.  For a GSoC project, you have more time, so should be doing a design that includes globs.

We already have code in coala core for how to handle `files` which have no results, and possibly also similar handling for each glob which has no results.

A very simple design is to introduce a setting in coala like `require_files_not_empty = true` and `require_files_for_each_glob = true` to select what should happen if there are globs missing or the entire `files` evaluates to 0 files.

Reply:
>A very simple design is to introduce a setting in coala like require_files_not_empty = true and require_files_for_each_glob = true to select what should happen if there are globs missing or the entire files evaluates to 0 files.

This would be quite different. If you want I'll get it changed to this.

Reply:
coala already emits a warning if any of the globs or files in `files` don't get any match. So, I guess we don't need any extra settings like these(?).